### PR TITLE
Version 214

### PR DESCRIPTION
**Pull request by vinniefalco**
* Handler binders use the associated allocator
* Add detail::bind_continuation
* Rewrite the echo-op example

### CODE REVIEW

File: example/echo-op/echo_op.cpp [hunk 1]
```diff
@@ -14,276 +14,324 @@
...
+
+                        // Keep looping until we get the newline
+                        continue;
+                    }
+
+                    // We have our newline, so send the first `pos` bytes of the
+                    // buffers. The function `buffers_prefix` returns the front part
+                    // of the buffers we want.
+
+                    yield net::async_write(stream_,
```
Comment (djarek) [inline example/echo-op/echo_op.cpp]: Can't this use `net::async_write(stream_, buffer.data(), net::transfer_exactly(pos), std::move(*this)`?

File: example/echo-op/echo_op.cpp [hunk 2]
```diff
@@ -14,276 +14,324 @@
...
+
+                    // The loop terminates here, and we will either deliver a
+                    // successful result or an error to the caller's completion handler.
+
+                    break;
+                }
+
+                // When a composed operation completes immediately, it must not
+                // directly invoke the completion handler otherwise it could
+                // lead to unfairness, starvation, or stack overflow. Therefore,
```
Comment (djarek) [inline example/echo-op/echo_op.cpp]: I'd just scare people that it's UB. If we're dealing with coroutines, calling the handler directly would resume a coroutine that is running.

File: example/echo-op/echo_op.cpp [hunk 3]
```diff
@@ -14,276 +14,324 @@
...
+                // if cont == false (meaning, that the call stack still includes
+                // the frame of the initiating function) then use `net::post`
+                // to cause us to be called again after the initiating function
+                // returns. The function `bind_handler` works similarly to
+                // `std::bind`, allowing bound arguments to be passed to our
+                // completion handler during the dispatch, but also takes care
+                // of forwarding the allocator and executor associations in
```
Comment (djarek) [inline example/echo-op/echo_op.cpp]: "associations" => "customization points"

File: test/beast/core/test_executor.hpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+    void
+    on_work_started() const noexcept
+    {
+        ++sp_->info_.work;
+    }
+
+    void
+    on_work_finished() const noexcept
+    {
```
Comment (djarek) [inline test/beast/core/test_executor.hpp]: Should this decrement the work count?

Comment (vinniefalco) [reply test/beast/core/test_executor.hpp]: No, the purpose of the counter is just to know that it was called

File: example/echo-op/echo_op.cpp (hunk 2 above)
Comment (vinniefalco) [reply example/echo-op/echo_op.cpp]: I copied Chris' wording from the executors paper

File: example/echo-op/echo_op.cpp (hunk 1 above)
Comment (vinniefalco) [reply example/echo-op/echo_op.cpp]: In theory yes, but its nice to see as many beast functions as possible in one example to show their use

**Pull request merged**